---
title: "Tutorial"
output: html_document
date: "2025-02-01"
---

This is a tutorial for using `multideconv` and explain how to use the main functions of the pipeline to deconvolve your RNAseq data. Details of the parameters of each function can be obtained from `cell_deconvolution.R`

Set up environment
```{r setup, include=FALSE}
source("cell_deconvolution.R") #Load functions and packages
set.seed(123) #For reproducibility
```

Load data
```{r}
raw.counts = read.csv("data/raw_counts_Mariathasan.csv", row.names = 1) 
```

Perform cell type deconvolution using 5 methods (quanTIseq, DeconRNASeq, CibersortX, EpiDISH, DWLS) and 10 signatures (check `signatures/` folder). The function takes as input either raw counts or TPM normalized counts (genes as SYMBOL) and it will ask for your credentials to run CBSX (see README for more information). It will save automatically the deconvolution matrix in the `Results/` directory.
```{r}
deconv = compute.deconvolution(raw.counts, normalized = T, credentials.mail = "xxxx", credentials.token = "xxxxxx", file_name = "Tutorial")
```

If you don't want to run a specific method or used a specific signature, you can exclude them using the following:
```{r}
#The argument methods = () specify which methods you want to run 
deconv = compute.deconvolution(raw.counts, normalized = T, credentials.mail = "xxxx", credentials.token = "xxxxxx", methods = c("Quantiseq", "MCP", "XCell", "DWLS"), file_name = "Tutorial")
#The argument signatures_exclude specify which signature you want to exclude, this should be the name of the signature file
deconv = compute.deconvolution(raw.counts, normalized = T, credentials.mail = "xxxx", credentials.token = "xxxxxx", signatures_exclude = "BPRNACan", file_name = "Tutorial")
```

For speeding up the running we have implemented parallelization jobs to perform deconvolution using variable signatures. You can do this by setting the parameter doParallel = T and specifying the number of workers your computer has (be careful to put a number corresponding to your computer characteristics!)
```{r}
deconv = compute.deconvolution(raw.counts, normalized = T, credentials.mail = "xxxx", credentials.token = "xxxxxx", file_name = "Tutorial", doParallel = T, workers = 3)
```

We have also implemented the option to use second-generation methods to deconvolve your data. These methods use as input single cell data to learn from it and deconvolve the bulk RNAseq. If you would like to use it, specify it in the parameters using the following.
```{r}
deconv = compute.deconvolution(raw.counts, normalized = T, credentials.mail = "xxxx", credentials.token = "xxxxxx", sc_deconv = T, sc_matrix = sc.object, cell_annotations = cell_labels, cell_samples = bath_ids, name_sc_signature = "Tutorial", file_name = "Tutorial")

#sc_deconv: Boolean value specifying if second-generation methods need to be run
#sc_matrix: Normalized gene expression matrix with genes as rows and cell as columns from the scRNAseq object
#cell_annotations: Vector corresponding with the cell annotations 
#cell_samples: Vector corresponding with the samples_ids of each cell
#name_sc_signature: Name you want to give to your signature
```

Deconvolution analysis corresponds to applying a subgrouping algorithm to reduce the dimensionality and remove the noise of your cell features. It returns cell type subgroups with grouped methods-signatures features correlated across your samples. It saves the subgroups information and the deconvolve matrix with the subgroups scores as .csv files in the `Results/` directory.
```{r}
dt = compute.deconvolution.analysis(deconvolution = deconv, corr = 0.7, seed = 123, file_name = "Tutorial") 

#deconvolution: A matrix with unprocessed cell deconvolution results
#corr: A numeric value with the minimum correlation allowed to group cell deconvolution features
#zero: A numeric value with the maximum proportion of zeros values allowed the deconvolution features to have. Features with higher number of zeros across samples (>zero) will be discarded. Default is 0.9
#high_corr: A numeric value with the threshold for pairwise correlation. Pair of features correlated with more than this threshold are classify as 'high correlated' and choose randomly one of them. Default is 0.9
#seed: A numeric value to specificy the seed. This ensures reproducibility during the choice step of high correlated features.
#return: Boolean value to whether return and saved the plot and csv files of deconvolution generated during the run inside the Results/ directory.
```


Deconvolution only first-generation
```{r}
deconv = compute.deconvolution(bulk_data, normalized = F, credentials.mail = "marcelo.hurtado@inserm.fr", credentials.token = "734212f6ad77fc4eea2bdb502792f294", file_name = "Pseudobulk", doParallel = T, workers = 4)
```

Deconvolve only second-generation methods
```{r}
deconv = compute_sc_deconvolution_methods(bulk_data, normalized = F, metacells@assays$RNA@counts, metacells@meta.data, "annotated_ct", 
                                          "sample", name_object = "Vanderbilt", n_cores = 6, return = T, file_name = "Vanderbilt")
```

Deconvolution first generation + create signatures 
```{r}
deconv = compute.deconvolution(bulk_data[,1:5], normalized = F, credentials.mail = "marcelo.hurtado@inserm.fr", credentials.token = "734212f6ad77fc4eea2bdb502792f294",
                               file_name = "Tutorial", create_signature = T,  sc_matrix = metacells$Counts, sc_metadata = metacells$Metadata, 
                               cell_label = "annotated_ct", sample_label = "sample", cell_markers = cell_type_hvg, name_sc_signature = "Vanderbilt", workers = 1)
```

Deconvolution all + create signatures (No recommended) 
```{r}
deconv = compute.deconvolution(bulk_data[,1:5], normalized = F, credentials.mail = "marcelo.hurtado@inserm.fr", credentials.token = "734212f6ad77fc4eea2bdb502792f294",
                               file_name = "Tutorial", create_signature = T,  sc_matrix = metacells$Counts, sc_metadata = metacells$Metadata, sc_deconv = T,
                               cell_label = "annotated_ct", sample_label = "sample", cell_markers = cell_type_hvg, name_sc_signature = "Vanderbilt", workers = 1)
```
